rllib/utils/exploration/parameter_noise.py

```diff
@@ -0,0 +1,386 @@
+        # corresponds to one Model variable and holding the Gaussian noise to
+        # be added to that variable (weight).
+        self.noise = []
+        for var in self.model_variables:
+            self.noise.append(
+                get_variable(
+                    np.zeros(var.shape, dtype=np.float32),
+                    framework=self.framework,
+                    tf_name=var.name.split(":")[0] + "_noisy"))
+
+        # tf-specific ops to sample, assign and remove noise.
+        if self.framework == "tf" and not tf.executing_eagerly():
+            self.tf_sample_new_noise_op = \
+                self._tf_sample_new_noise_op()
+            self.tf_add_stored_noise_op = \
+                self._tf_add_stored_noise_op()
+            self.tf_remove_noise_op = \
+                self._tf_remove_noise_op()
+            # Create convenience sample+add op for tf.
+            with tf.control_dependencies([self.tf_sample_new_noise_op]):
+                add_op = self._tf_add_stored_noise_op()
+            with tf.control_dependencies([add_op]):
+                self.tf_sample_new_noise_and_add_op = tf.no_op()
+
+        # Whether the Model's weights currently have noise added or not.
+        self.weights_are_currently_noisy = False
+
+        # Auto-detection of underlying exploration functionality.
+        if sub_exploration is None:
+            # For discrete action spaces, use an underlying EpsilonGreedy with
+            # a special schedule.
+            if isinstance(self.action_space, Discrete):
+                sub_exploration = {
+                    "type": "EpsilonGreedy",
+                    "epsilon_schedule": {
+                        "type": "PiecewiseSchedule",
+                        # Step function (see [2]).
+                        "endpoints": [(0, 1.0), (random_timesteps + 1, 1.0),
+                                      (random_timesteps + 2, 0.01)],
+                        "outside_value": 0.01
+                    }
+                }
+            # TODO(sven): Implement for any action space.
+            else:
+                raise NotImplementedError
+
+        self.sub_exploration = from_config(
+            Exploration,
+            sub_exploration,
+            framework=self.framework,
+            action_space=self.action_space,
+            **kwargs)
+
+        # Store the default setting for `explore`.
+        self.default_explore = policy_config["explore"]
+        # Whether we need to call `self._delayed_on_episode_start` before
+        # the forward pass.
+        self.episode_started = False
+
+    @override(Exploration)
+    def before_forward_pass(self,
+                            obs_batch,
+                            *,
+                            state_batches=None,
+                            seq_lens=None,
+                            prev_action_batch=None,
+                            prev_reward_batch=None,
```

Comment:
Since we don't need these flags (and I can't think of a common use case for them beyond param noise), can you remove these optional args for now? This is an internal-ish developer API so it is ok to make changes to add more fields as needed in the future.

Reply:
done.